app/components/model-renderer.jsx

```diff
@@ -0,0 +1,148 @@
+import React from 'react';
+import { Markdown } from 'markdownz';
+import { Model } from './modelling';
+import alert from '../lib/alert';
+
+// function to check whether webgl is available
+function webGLCompatibilityTest() {
+  try {
+    return !!window.WebGLRenderingContext && !!document.createElement('canvas').getContext('experimental-webgl');
+  } catch (e) {
+    return false;
+  }
+}
+
+class ModelRenderer extends React.Component {
+  constructor(props) {
+    super(props);
+    this.getDifference = this.getDifference.bind(this);
+    this.setTexture = this.setTexture.bind(this);
+
+    if (this.props.subject.locations.length < 2) {
+      const im0 = this.props.subject.locations[0];
+      this.props.subject.locations[1] = Object.assign({}, im0);
+    }
+    this.state = {
+      imgUrl: Object.values(this.props.subject.locations[0])[0]
+    };
+  }
+  componentDidMount() {
+    if (!webGLCompatibilityTest()) {
+      // TODO: Render an error message to the screen
+      /* eslint-disable no-console */
+      alert(
+        (resolve, reject) =>
+          (
+            <div className="content-container">
+              <Markdown className="classification-task-help">
+                WebGL is required for this project, please try again using an updated browser.
+              </Markdown>
+              <button className="standard-button" onClick={reject}>Close</button>
+            </div>
+          )
+      );
+      /* eslint-enable no-console */
+      return;
+    }
+    // component has mounted, initialise the regl canvas
+    this.model = new Model(this.canvas, this.props.subject.metadata);
+  }
+  componentWillReceiveProps() {
+    // component has updated with new props, set the new render functions
+    // check if a model has been assigned
+    if (this.model !== null) {
+      // if so, update the render function list to the new annotation
+      window.requestAnimationFrame(this.getDifference);
+    }
+  }
+
+  // don't want to re-render whenever props update
+  shouldComponentUpdate() {
+    if (this.state.imgUrl !== Object.values(this.props.subject.locations[0])[0]) {
+      // we have a new subject. Refresh the model calculator
+      const im0 = this.props.subject.locations[0];
+      if (this.props.subject.locations.length < 2) {
+        this.props.subject.locations[1] = Object.assign({}, im0);
+      }
+      this.model.kill();
+      const metadata = this.props.subject.metadata;
+      this.model = new Model(this.canvas, metadata);
+      this.setState({ imgUrl: Object.values(im0)[0] });
```

Comment:
`shouldComponentUpdate` should only contain logic around returning `true` or `false`. This other code should go in `componentWillUpdate` (code called *before* an update happens).